sunscreen_compiler_common/src/context.rs

```diff
@@ -444,5 +463,48 @@ where
      */
     pub fn add_edge(&mut self, from: NodeIndex, to: NodeIndex, edge: EdgeInfo) {
         self.graph.add_edge(from, to, edge);
+
+        #[cfg(feature = "debugger")]
+        {
+            let group_lookup = &mut self.graph.metadata.group_lookup;
+            let from_index: u64 = from.index().try_into().unwrap();
+            let to_index: u64 = to.index().try_into().unwrap();
+
+            let group_from_id = &self.graph.graph.node_weight(from).unwrap().group_id;
+            let group_to_id = &self.graph.graph.node_weight(to).unwrap().group_id;
+
+            if let Some(group_from) = group_lookup.id_data_lookup.get_mut(group_from_id) {
+                if !group_from.node_ids.contains(&to_index) {
+                    group_from.node_outputs.push(to_index);
+                }
+            }
+
+            if let Some(group_to) = group_lookup.id_data_lookup.get_mut(group_to_id) {
+                if !group_to.node_ids.contains(&from_index) {
+                    group_to.node_inputs.push(from_index);
+                }
+            }
+        }
+    }
+
+    #[cfg(feature = "debugger")]
+    /**
+     *
+     */
+    pub fn push_group(&mut self, name: String, source: String) {
+        let parent = *self.group_stack.last().unwrap_or(&0);
+        let group = Group::new(
+            self.next_group_id,
+            format!("{}_{}", name, self.next_group_id),
+            parent,
+            source,
+        );
+        self.group_stack.push(self.next_group_id);
+        self.graph
+            .metadata
+            .group_lookup
+            .id_data_lookup
+            .insert(self.next_group_id, group);
```

Comment:
I see, you're using parent to chase up the stack and find the root rather than using the entire group stack as keys... I think that can work.